generator/app/models/generator/ScalaDatatype.scala

```diff
@@ -146,68 +166,28 @@ object ScalaPrimitive {
 
 }
 
-sealed trait ScalaDatatype {
-  def nilValue: String
-
-  def primitive: ScalaPrimitive
-
-  def name: String
-
-  def clientDefinition(
-    originalVarName: String,
-    optional: Boolean
-  ): String = {
-    val varName = ScalaUtil.quoteNameIfKeyword(originalVarName)
-    if (optional) {
-      s"$varName: _root_.scala.Option[$name] = None"
-    } else {
-      s"$varName: $name"
-    }
-  }
-
-  /**
-    * For collections, the server side optional definition uses the
-    * nil value - e.g. a List will be Nil (instead of None).
-    */
-  def serverDefinition(
-    originalVarName: String,
-    optional: Boolean
-  ): String = {
-    val varName = ScalaUtil.quoteNameIfKeyword(originalVarName)
-    if (optional) {
-      s"$varName: $name = " + nilValue
-    } else {
-      s"$varName: $name"
-    }
-  }
-
-}
-
 object ScalaDatatype {
+  sealed abstract class Container(inner: ScalaDatatype) extends ScalaDatatype
 
-  case class List(primitive: ScalaPrimitive) extends ScalaDatatype {
-    override def nilValue = "Nil"
+  case class List(primitive: ScalaPrimitive) extends Container(primitive) {
     override def name = s"Seq[${primitive.fullName}]"
   }
 
-  case class Map(primitive: ScalaPrimitive) extends ScalaDatatype {
-    override def nilValue = "Map.empty"
+  case class Map(primitive: ScalaPrimitive) extends Container(primitive) {
     override def name = s"Map[String, ${primitive.fullName}]"
   }
 
-  case class Singleton(primitive: ScalaPrimitive) extends ScalaDatatype {
-    override def nilValue = "None"
-    override def name = primitive.fullName
+  case class Option(datatype: ScalaDatatype) extends Container(datatype) {
+    override def name = s"_root_.scala.Option[${datatype.name}]"
 
-    /**
-      * For singleton, server definition matches client definition
-      * wrapping the type in an Option always.
-      */
-    override def serverDefinition(
+    override def definition(
       originalVarName: String,
-      optional: Boolean
-    ): String = clientDefinition(originalVarName, optional)
-
+      default: scala.Option[String]
+    ): String = {
+      require(default.isEmpty, s"no defaults allowed on options: ${default}")
```

Comment:
This implies changing the validation of uploaded api.json files to disallow setting `"required": false` while also setting a default. In part this PR proposes to have defaults be sent by the client, instead of picked on the server. This implies tha t a field with a default value is actually required, since if unset by the user, it will be defaulted in the generated client.